ADAL/src/ui/ADWebAuthController.m

```diff
@@ -63,538 +64,112 @@
+    return result;
 }
+#endif // TARGET_OS_IPHONE
+@end
 
-- (BOOL)endWebAuthenticationWithError:(ADAuthenticationError*) error
-                                orURL:(NSURL*)endURL
-{
-    if (!_authenticationViewController)
-    {
-        return NO;
-    }
-    
-    [_authenticationViewController stop:^{[self dispatchCompletionBlock:error URL:endURL];}];
-    _authenticationViewController = nil;
-    
-    return YES;
-}
+#pragma mark - Private Methods
 
-- (void)onStartActivityIndicator:(id)sender
-{
-    (void)sender;
-    
-    if (_loading)
-    {
-        [_authenticationViewController startSpinner];
-    }
-    _spinnerTimer = nil;
-}
+@implementation ADWebAuthController (Internal)
 
-- (void)stopSpinner
++ (void)registerWebAuthNotifications
 {
-    if (!_loading)
-    {
-        return;
-    }
-    
-    _loading = NO;
-    if (_spinnerTimer)
-    {
-        [_spinnerTimer invalidate];
-        _spinnerTimer = nil;
-    }
-    
-    [_authenticationViewController stopSpinner];
+    MSIDNotifications.webAuthDidCompleteNotificationName = ADWebAuthDidCompleteNotification;
+    MSIDNotifications.webAuthDidFailNotificationName = ADWebAuthDidFailNotification;
+    MSIDNotifications.webAuthDidStartLoadNotificationName = ADWebAuthDidStartLoadNotification;
+    MSIDNotifications.webAuthDidFinishLoadNotificationName = ADWebAuthDidFinishLoadNotification;
 }
 
-#pragma mark - ADWebAuthDelegate
-
-- (void)webAuthDidStartLoad:(NSURL*)url
++ (void)startWithRequest:(ADRequestParameters *)requestParams
+          promptBehavior:(ADPromptBehavior)promptBehavior
+                 context:(ADAuthenticationContext *)context
+              completion:(MSIDWebviewAuthCompletionHandler)completionHandler
 {
-    if (!_loading)
-    {
-        _loading = YES;
-        if (_spinnerTimer)
-        {
-            [_spinnerTimer invalidate];
-        }
-        _spinnerTimer = [NSTimer scheduledTimerWithTimeInterval:2.0
-                                                         target:self
-                                                       selector:@selector(onStartActivityIndicator:)
-                                                       userInfo:nil
-                                                        repeats:NO];
-        [_spinnerTimer setTolerance:0.3];
-    }
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        [self registerWebAuthNotifications];
+    });
     
-    [[NSNotificationCenter defaultCenter] postNotificationName:ADWebAuthDidStartLoadNotification object:self userInfo:url ? @{ @"url" : url } : nil];
-}
-
-- (void)webAuthDidFinishLoad:(NSURL*)url
-{
-    MSID_LOG_VERBOSE(_requestParams, @"-webAuthDidFinishLoad host: %@", [ADAuthorityUtils isKnownHost:url] ? url.host : @"unknown host");
-    MSID_LOG_VERBOSE_PII(_requestParams, @"-webAuthDidFinishLoad host: %@", url.host);
+    NSString *authorityWithOAuthSuffix = [NSString stringWithFormat:@"%@%@", context.authority, MSID_OAUTH2_AUTHORIZE_SUFFIX];
```

Comment:
nit: shouldn't we have a common core or ADAL utility to compose this url?

Reply:
It is being addressed in https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/130 , 
I will at least add some comment on this to update when the above PR is merged.